

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Data Integration (Batch correction) &#8212; Bioinfozone - Easy way to learn Bioinformatics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'part1/chapter5/Data Integration';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Clustering" href="../chapter6/Clustering.html" />
    <link rel="prev" title="Dimensionality Reduction" href="../chapter4/Dimensionality%20Reduction.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../intro.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/Bioinfozone logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/Bioinfozone logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Bioinfo</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../part0/chapter1/Introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Single cell RNAseq Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chapter1/Set-Up.html">Set-Up System :</a></li>




<li class="toctree-l1"><a class="reference internal" href="../chapter2/Data.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter3/Pre-processing.html">Pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter4/Dimensionality%20Reduction.html">Dimensionality Reduction</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Data Integration (Batch correction)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter6/Clustering.html">Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter7/Differential%20Gene%20Expression.html">Differential Gene Expression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Single Cell Multiomics Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../part2/chapter1/Introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Cloud Computing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter1/Introduction.html">Introduction</a></li>


<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter2/SetupGCP.html">Creating New VM instance ( Like setting NEW linux machine) :</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter3/setupVNCserver.html">install VNC server</a></li>




<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter4/STARAlignment.html">RNAseq Alignment</a></li>





</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">



<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Data Integration (Batch correction)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#check-batch-correction-needed">Check Batch correction Needed ?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-correction">Batch Correction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scvi-data-integration-batch-correction">scvi Data integration (Batch correction)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bbknn-integration">BBKNN Integration</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scanoama">scanoama</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="data-integration-batch-correction">
<h1>Data Integration (Batch correction)<a class="headerlink" href="#data-integration-batch-correction" title="Permalink to this heading">#</a></h1>
<p>Batch effects are changes in gene expression due to batches arise by different handling conditions such as , library depth, machines, Days, Stress management during extraction, even samples etc.</p>
<p>But selecting batch and label key is important . according to requirement of keeping batch</p>
<p>general, one can say that Harmony and Seurat consistently perform well for simple batch correction tasks, and scVI, scGen, scANVI, and Scanorama perform well for more complex data integration tasks.</p>
<p><strong>Note</strong></p>
<p>Previous Feature selection used Normalized , scaled data . But for Batch correction it is important to use RawData and find variable genes based on batch (Not on whole data)</p>
<p>It is important to use Rawdata</p>
<ul class="simple">
<li><p>RawData as adata2</p></li>
<li><p>We also use Filted DM data as adata</p></li>
</ul>
<section id="check-batch-correction-needed">
<h2>Check Batch correction Needed ?<a class="headerlink" href="#check-batch-correction-needed" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Load DM reduced filtered data:</span>

<span class="n">adata_QCNFSDM</span> <span class="o">=</span>  <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;/content/drive/MyDrive/scRNA_using_Python/Objects/sc_QCNFSDM_covid.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata_QCNFSDM</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;, &#39;highly_variable&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;doublet_colors&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;tsne&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_tsne&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;APR_counts&#39;, &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Load Raw Data :</span>

<span class="n">adata_raw</span> <span class="o">=</span>  <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;/content/drive/MyDrive/scRNA_using_Python/Objects/adata_raw_covid.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata_raw</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="o">.</span><span class="n">X</span>
<span class="n">adata_raw</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 33538
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-2d21dce5-a6b4-40f0-8547-6e4c606f91ed">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>type</th>
      <th>sample</th>
      <th>batch</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AGGGTCCCATGACCCG-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>TACCCACAGCGGGTTA-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>CCCAACTTCATATGGC-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>TCAAGTGTCCGAACGC-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>ATTCCTAGTGACTGTT-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CGCATAATCTTACGGA-14-5</th>
      <td>Ctrl</td>
      <td>ctrl_14</td>
      <td>5</td>
    </tr>
    <tr>
      <th>GAGGCCTTCTCCTGCA-14-5</th>
      <td>Ctrl</td>
      <td>ctrl_14</td>
      <td>5</td>
    </tr>
    <tr>
      <th>CCCTAACAGTTTCTTC-14-5</th>
      <td>Ctrl</td>
      <td>ctrl_14</td>
      <td>5</td>
    </tr>
    <tr>
      <th>GGGATGATCAAGCTTG-14-5</th>
      <td>Ctrl</td>
      <td>ctrl_14</td>
      <td>5</td>
    </tr>
    <tr>
      <th>CAATGACCACTGCATA-14-5</th>
      <td>Ctrl</td>
      <td>ctrl_14</td>
      <td>5</td>
    </tr>
  </tbody>
</table>
<p>9000 rows × 3 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-2d21dce5-a6b4-40f0-8547-6e4c606f91ed')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-2d21dce5-a6b4-40f0-8547-6e4c606f91ed button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-2d21dce5-a6b4-40f0-8547-6e4c606f91ed');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  </div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Assign key for batch and label </span>
<span class="n">label_key</span> <span class="o">=</span> <span class="s2">&quot;type&quot;</span>
<span class="n">batch_key</span> <span class="o">=</span> <span class="s2">&quot;sample&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>covid_1     1500
covid_15    1500
covid_17    1500
ctrl_5      1500
ctrl_13     1500
ctrl_14     1500
Name: sample, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_raw</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene Expression    33538
Name: feature_types, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title filtering to make sure we have no features with zero counts</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">adata_raw</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 21830
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>we also need to re-normalise the data. Here we just normalise using global scaling by the total counts per cell.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title simple normalize log1p</span>

<span class="n">adata_raw</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="n">adata_raw</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">adata_raw</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 21830
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;
    uns: &#39;log1p&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="n">adata_raw</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 21830
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Batch correction needed ?</span>

<span class="n">adata_raw</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">batch_key</span> <span class="o">+</span> <span class="s2">&quot;_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;#1b9e77&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#d95f02&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#7570b3&quot;</span><span class="p">,</span>
<span class="p">]</span>  <span class="c1"># Set custom colours for batches</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/4922e68bcece773e1e29b5c5ea4159f4ccec6ee7cd3f533bd35e86776668b9ee.png" src="../../_images/4922e68bcece773e1e29b5c5ea4159f4ccec6ee7cd3f533bd35e86776668b9ee.png" />
</div>
</div>
<p>We need batch correction according to sample batch</p>
</section>
<section id="batch-correction">
<h2>Batch Correction<a class="headerlink" href="#batch-correction" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Feature selection wrt Batch :</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
    <span class="n">adata_raw</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span>
<span class="p">)</span>
<span class="n">adata_raw</span>
<span class="n">adata_raw</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-a756f2fb-3e44-487a-a4fb-18d839c4e3a9">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
      <th>feature_types</th>
      <th>genome</th>
      <th>n_cells</th>
      <th>highly_variable</th>
      <th>means</th>
      <th>dispersions</th>
      <th>dispersions_norm</th>
      <th>highly_variable_nbatches</th>
      <th>highly_variable_intersection</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AL627309.1</th>
      <td>ENSG00000238009</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>25</td>
      <td>False</td>
      <td>0.001361</td>
      <td>0.555600</td>
      <td>-0.151289</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AL627309.3</th>
      <td>ENSG00000239945</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>1</td>
      <td>False</td>
      <td>0.000041</td>
      <td>0.061465</td>
      <td>-0.002369</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AL669831.5</th>
      <td>ENSG00000237491</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>461</td>
      <td>False</td>
      <td>0.026669</td>
      <td>0.737217</td>
      <td>-0.089099</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>FAM87B</th>
      <td>ENSG00000177757</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>7</td>
      <td>False</td>
      <td>0.000366</td>
      <td>0.368892</td>
      <td>0.569010</td>
      <td>1</td>
      <td>False</td>
    </tr>
    <tr>
      <th>LINC00115</th>
      <td>ENSG00000225880</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>208</td>
      <td>False</td>
      <td>0.011747</td>
      <td>0.654689</td>
      <td>-0.411236</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>AC007325.4</th>
      <td>ENSG00000278817</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>42</td>
      <td>False</td>
      <td>0.002123</td>
      <td>0.618409</td>
      <td>-0.435802</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AL354822.1</th>
      <td>ENSG00000278384</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>43</td>
      <td>False</td>
      <td>0.003140</td>
      <td>0.822280</td>
      <td>1.033842</td>
      <td>2</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AC004556.1</th>
      <td>ENSG00000276345</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>400</td>
      <td>False</td>
      <td>0.022188</td>
      <td>0.670790</td>
      <td>0.646585</td>
      <td>1</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AC233755.1</th>
      <td>ENSG00000275063</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>4</td>
      <td>False</td>
      <td>0.000313</td>
      <td>0.385842</td>
      <td>0.627800</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AC240274.1</th>
      <td>ENSG00000271254</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>61</td>
      <td>False</td>
      <td>0.003973</td>
      <td>0.839225</td>
      <td>0.555512</td>
      <td>1</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>21830 rows × 10 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-a756f2fb-3e44-487a-a4fb-18d839c4e3a9')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-a756f2fb-3e44-487a-a4fb-18d839c4e3a9 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-a756f2fb-3e44-487a-a4fb-18d839c4e3a9');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  </div></div>
</div>
<ul class="simple">
<li><p>highly_variable_nbatches - The number of batches where each gene was found to be highly variable</p></li>
<li><p>highly_variable_intersection - Whether each gene was highly variable in every batch</p></li>
<li><p>highly_variable - Whether each gene was selected as highly variable after combining the results from each batch</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title No of Batches eaach gene</span>

<span class="n">n_batches</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable_nbatches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">n_batches</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">)</span>
<span class="n">n_batches</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0    14794
1     4371
2     1440
3      618
4      286
5      175
6      146
Name: highly_variable_nbatches, dtype: int64
</pre></div>
</div>
<img alt="../../_images/cef0151bd6761e2b124f0ee8150dec2e2fb2153cc73f0dc31aaff25ae4e4eb9b.png" src="../../_images/cef0151bd6761e2b124f0ee8150dec2e2fb2153cc73f0dc31aaff25ae4e4eb9b.png" />
</div>
</div>
<p>most genes are not highly variable.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title create object to use for Integration</span>

<span class="n">adata_hvg</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="p">[:,</span> <span class="n">adata_raw</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_hvg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<section id="scvi-data-integration-batch-correction">
<h3>scvi Data integration (Batch correction)<a class="headerlink" href="#scvi-data-integration-batch-correction" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scvi</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>prepare our AnnData object. This step stores some information required by scVI such as which expression matrix to use and what the batch key is.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">)</span>
<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">model_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">SCVI Model with the following params: 
n_hidden: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">128</span>, n_latent: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">10</span>, n_layers: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>, dropout_rate: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.1</span>, dispersion: gene, gene_likelihood: zinb, 
latent_distribution: normal
Training status: Not Trained
Model's adata is minified?: <span style="color: #ff0000; text-decoration-color: #ff0000; font-style: italic">False</span>
</pre>
</div><div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span><span class="o">.</span><span class="n">view_anndata_setup</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Anndata setup with scvi-tools version <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.20</span>.<span style="color: #008080; text-decoration-color: #008080; font-weight: bold">3</span>.
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Setup via `SCVI.setup_anndata` with arguments:
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">{</span>
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'layer'</span>: <span style="color: #008000; text-decoration-color: #008000">'counts'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'batch_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'sample'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'labels_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'size_factor_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'categorical_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'continuous_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>
<span style="font-weight: bold">}</span>
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">         Summary Statistics         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃<span style="font-weight: bold">     Summary Stat Key     </span>┃<span style="font-weight: bold"> Value </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_batch          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   6   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_cells          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 9000  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_categorical_covs </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_continuous_covs  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_labels         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   1   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">          n_vars          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 2000  </span>│
└──────────────────────────┴───────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">               Data Registry                </span>
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Registry Key </span>┃<span style="font-weight: bold">    scvi-tools Location    </span>┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">      X       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  adata.layers['counts']   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    batch     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_batch']  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    labels    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_labels'] </span>│
└──────────────┴───────────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                   batch State Registry                   </span>
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">   Source Location   </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['sample'] </span>│<span style="color: #008000; text-decoration-color: #008000">  covid_1   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                     </span>│<span style="color: #008000; text-decoration-color: #008000">  covid_15  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                     </span>│<span style="color: #008000; text-decoration-color: #008000">  covid_17  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                     </span>│<span style="color: #008000; text-decoration-color: #008000">   ctrl_5   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          3          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                     </span>│<span style="color: #008000; text-decoration-color: #008000">  ctrl_13   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          4          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                     </span>│<span style="color: #008000; text-decoration-color: #008000">  ctrl_14   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          5          </span>│
└─────────────────────┴────────────┴─────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                     labels State Registry                      </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">      Source Location      </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['_scvi_labels'] </span>│<span style="color: #008000; text-decoration-color: #008000">     0      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
└───────────────────────────┴────────────┴─────────────────────┘
</pre>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">max_epochs_scvi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="nb">round</span><span class="p">((</span><span class="mi">20000</span> <span class="o">/</span> <span class="n">adata_scvi</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span> <span class="o">*</span> <span class="mi">400</span><span class="p">),</span> <span class="mi">400</span><span class="p">])</span>
<span class="n">max_epochs_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>400
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:IPU available: False, using: 0 IPUs
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [04:31&lt;00:00,  1.47it/s, loss=429, v_num=1]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=400` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [04:31&lt;00:00,  1.48it/s, loss=429, v_num=1]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scvi</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_scvi</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>

<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_scVI&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_scVI&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/3af4b180b936ad995bfb355a47923e31a45e67fe31fc26d4fac7e5ad3ae2c896.png" src="../../_images/3af4b180b936ad995bfb355a47923e31a45e67fe31fc26d4fac7e5ad3ae2c896.png" />
</div>
</div>
<p>Much Better . But we didn’t get corrected metrix</p>
</section>
<section id="bbknn-integration">
<h3>BBKNN Integration<a class="headerlink" href="#bbknn-integration" title="Permalink to this heading">#</a></h3>
<p>An important parameter for BBKNN is the number of neighbors per batch. A suggested heuristic for this is to use 25 if there are more than 100,000 cells or the default of 3 if there are fewer than 100,000.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">neighbors_within_batch</span> <span class="o">=</span> <span class="mi">25</span> <span class="k">if</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">n_obs</span> <span class="o">&gt;</span> <span class="mi">100000</span> <span class="k">else</span> <span class="mi">3</span>
<span class="n">neighbors_within_batch</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>3
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bbknn</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_bbknn</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_bbknn</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bbknn</span><span class="o">.</span><span class="n">bbknn</span><span class="p">(</span>
    <span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">,</span> <span class="n">neighbors_within_batch</span><span class="o">=</span><span class="n">neighbors_within_batch</span>
<span class="p">)</span>
<span class="n">adata_bbknn</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/4a7a6fb8c66818ac3c0785e7f8ddfdb8e9dc408d72caf5937821c6238a552740.png" src="../../_images/4a7a6fb8c66818ac3c0785e7f8ddfdb8e9dc408d72caf5937821c6238a552740.png" />
</div>
</div>
<p>Much Better</p>
</section>
<section id="scanoama">
<h3>scanoama<a class="headerlink" href="#scanoama" title="Permalink to this heading">#</a></h3>
<p>To run Scanorama, you need to install python-annoy (already included in conda environment) and scanorama with pip. We can run scanorama to get a corrected matrix with the correct function, or to just get the data projected onto a new common dimension with the function integrate. Or both with the correct_scanpy and setting return_dimred=True. For now, run with just integration.</p>
<p>First we need to create individual AnnData objects from each of the datasets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scanorama</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">adata_scanorama</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># split per batch into new objects.</span>
<span class="n">batches</span> <span class="o">=</span> <span class="n">adata_scanorama</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">alldata</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">batch</span> <span class="ow">in</span> <span class="n">batches</span><span class="p">:</span>
    <span class="n">alldata</span><span class="p">[</span><span class="n">batch</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_scanorama</span><span class="p">[</span><span class="n">adata_scanorama</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">batch</span><span class="p">,]</span>

<span class="n">alldata</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;covid_1&#39;: View of AnnData object with n_obs × n_vars = 1500 × 2000
     obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
     var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
     uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     layers: &#39;counts&#39;, &#39;logcounts&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 &#39;covid_15&#39;: View of AnnData object with n_obs × n_vars = 1500 × 2000
     obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
     var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
     uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     layers: &#39;counts&#39;, &#39;logcounts&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 &#39;covid_17&#39;: View of AnnData object with n_obs × n_vars = 1500 × 2000
     obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
     var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
     uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     layers: &#39;counts&#39;, &#39;logcounts&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 &#39;ctrl_5&#39;: View of AnnData object with n_obs × n_vars = 1500 × 2000
     obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
     var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
     uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     layers: &#39;counts&#39;, &#39;logcounts&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 &#39;ctrl_13&#39;: View of AnnData object with n_obs × n_vars = 1500 × 2000
     obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
     var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
     uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     layers: &#39;counts&#39;, &#39;logcounts&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 &#39;ctrl_14&#39;: View of AnnData object with n_obs × n_vars = 1500 × 2000
     obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
     var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
     uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     layers: &#39;counts&#39;, &#39;logcounts&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scanorama</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable_nbatches&#39;</span><span class="p">]</span>

<span class="n">var_select</span> <span class="o">=</span> <span class="n">adata_scanorama</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable_nbatches&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span>
<span class="n">var_genes</span> <span class="o">=</span> <span class="n">var_select</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">var_select</span><span class="p">]</span>
<span class="nb">len</span><span class="p">(</span><span class="n">var_genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1225
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#subset the individual dataset to the variable genes we defined at the beginning</span>
<span class="n">alldata2</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">alldata</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
    <span class="n">alldata2</span><span class="p">[</span><span class="n">ds</span><span class="p">]</span> <span class="o">=</span> <span class="n">alldata</span><span class="p">[</span><span class="n">ds</span><span class="p">][:,</span><span class="n">var_genes</span><span class="p">]</span>

<span class="c1">#convert to list of AnnData objects</span>
<span class="n">adatas</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alldata2</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

<span class="c1"># run scanorama.integrate</span>
<span class="n">scanorama</span><span class="o">.</span><span class="n">integrate_scanpy</span><span class="p">(</span><span class="n">adatas</span><span class="p">,</span> <span class="n">dimred</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>covid_1
covid_15
covid_17
ctrl_5
ctrl_13
ctrl_14
Found 1225 genes among all datasets
[[0.         0.606      0.34933333 0.53266667 0.38866667 0.25733333]
 [0.         0.         0.63866667 0.478      0.26133333 0.19266667]
 [0.         0.         0.         0.52266667 0.20933333 0.20733333]
 [0.         0.         0.         0.         0.84133333 0.76466667]
 [0.         0.         0.         0.         0.         0.84133333]
 [0.         0.         0.         0.         0.         0.        ]]
Processing datasets (4, 5)
Processing datasets (3, 4)
Processing datasets (3, 5)
Processing datasets (1, 2)
Processing datasets (0, 1)
Processing datasets (0, 3)
Processing datasets (2, 3)
Processing datasets (1, 3)
Processing datasets (0, 4)
Processing datasets (0, 2)
Processing datasets (1, 4)
Processing datasets (0, 5)
Processing datasets (2, 4)
Processing datasets (2, 5)
Processing datasets (1, 5)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#scanorama adds the corrected matrix to adata.obsm in each of the datasets in adatas.</span>

<span class="n">adatas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_scanorama&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1500, 50)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get all the integrated matrices.</span>
<span class="n">scanorama_int</span> <span class="o">=</span> <span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_scanorama&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ad</span> <span class="ow">in</span> <span class="n">adatas</span><span class="p">]</span>

<span class="c1"># make into one matrix.</span>
<span class="n">all_s</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">scanorama_int</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">all_s</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="c1"># add to the AnnData object, create a new object first</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">adata_scanorama</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_sc</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;Scanorama&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_s</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(9000, 50)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9000 × 2000
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;Scanorama&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># tsne and umap</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">n_pcs</span> <span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span> <span class="o">=</span> <span class="s2">&quot;Scanorama&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">n_pcs</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span> <span class="o">=</span> <span class="s2">&quot;Scanorama&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span><span class="n">constrained_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Scanorama tsne&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Axes: title={&#39;center&#39;: &#39;Scanorama tsne&#39;}, xlabel=&#39;UMAP1&#39;, ylabel=&#39;UMAP2&#39;&gt;
</pre></div>
</div>
<img alt="../../_images/0abbea4e2240780144bb28021363ec5a90b5c6fe40c385a64136bed21e2bde13.png" src="../../_images/0abbea4e2240780144bb28021363ec5a90b5c6fe40c385a64136bed21e2bde13.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title save scanroma integrated data to file</span>
<span class="n">save_file</span> <span class="o">=</span> <span class="s1">&#39;Objects/sc_QCNFSDM_scanorama_corrected_covid.h5ad&#39;</span>
<span class="n">adata_sc</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">save_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./part1\chapter5"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../chapter4/Dimensionality%20Reduction.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Dimensionality Reduction</p>
      </div>
    </a>
    <a class="right-next"
       href="../chapter6/Clustering.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Clustering</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#check-batch-correction-needed">Check Batch correction Needed ?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-correction">Batch Correction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scvi-data-integration-batch-correction">scvi Data integration (Batch correction)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bbknn-integration">BBKNN Integration</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scanoama">scanoama</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Dr. Majeed Jamakhani
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>