

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Dimensionality Reduction &#8212; Bioinfozone - Easy way to learn Bioinformatics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'part1/chapter4/Dimensionality Reduction';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Data Integration (Batch correction)" href="../chapter5/Data%20Integration.html" />
    <link rel="prev" title="Pre-processing" href="../chapter3/Pre-processing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../intro.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/Bioinfozone logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/Bioinfozone logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Bioinfo</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../part0/chapter1/Introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Single cell RNAseq Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chapter1/Set-Up.html">Set-Up System :</a></li>




<li class="toctree-l1"><a class="reference internal" href="../chapter2/Data.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter3/Pre-processing.html">Pre-processing</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter5/Data%20Integration.html">Data Integration (Batch correction)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter6/Clustering.html">Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chapter7/Differential%20Gene%20Expression.html">Differential Gene Expression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Single Cell Multiomics Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../part2/chapter1/Introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Cloud Computing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter1/Introduction.html">Introduction</a></li>


<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter2/SetupGCP.html">Creating New VM instance ( Like setting NEW linux machine) :</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter3/setupVNCserver.html">install VNC server</a></li>




<li class="toctree-l1"><a class="reference internal" href="../../part3/chapter4/STARAlignment.html">RNAseq Alignment</a></li>





</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">



<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Dimensionality Reduction</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#normalization">Normalization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-selection">Feature Selection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-qc-fitered-normalized-feature-selected-object">Save QC fitered Normalized Feature selected object</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-tsne-umap">PCA,tSNE,UMAP</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-qc-filtered-feature-selected-dm-reduced-data-object">Save QC filtered feature selected , DM reduced data object</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="dimensionality-reduction">
<h1>Dimensionality Reduction<a class="headerlink" href="#dimensionality-reduction" title="Permalink to this heading">#</a></h1>
<section id="normalization">
<h2>Normalization<a class="headerlink" href="#normalization" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Load quality controlled data </span>

<span class="n">adata_QC</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span> <span class="s2">&quot;/content/drive/MyDrive/scRNA_using_Python/Objects/sc_qc_filtered_covid.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata_QC</span>

</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;ambient_counts&#39;, &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title check count distribution :</span>

<span class="n">p1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_QC</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/b96e6ddbe8462f6869991331e7c259172a7c532d303bb6dd952efaddd16957e7.png" src="../../_images/b96e6ddbe8462f6869991331e7c259172a7c532d303bb6dd952efaddd16957e7.png" />
</div>
</div>
<blockquote>
<div><p>Note :</p>
</div></blockquote>
<p>At this stage we have cleaned data according to filtered cells , corrected ambient RNA and doublet removed .</p>
<p>But data we had count Matrix (cells  X  Genes) , still variance due to sampling effect (cell capturing, library preparation, sequencing ) .</p>
<p>The difference we get between cells gene expression may be because of these sampling effects not by real difference . so we need to Normalize to correct for variance.</p>
<p>Normalize by 2 methods :</p>
<ol class="arabic simple">
<li><p>Log shifted Normalization</p></li>
</ol>
<ul class="simple">
<li><p>doesn’t prevent heterogenity</p></li>
<li><p>corrects for library depth</p></li>
</ul>
<ol class="arabic simple" start="2">
<li><p>Pearson residual Normalization (prevents heterogenity)</p></li>
</ol>
<ul class="simple">
<li><p>prevent heterogenity</p></li>
<li><p>corrects for library depth</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Method-1 : Log shifted method :</span>


<span class="n">scales_counts</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_QC</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># log1p transform</span>
<span class="n">adata_QC</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log1p_norm&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">scales_counts</span><span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">],</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">adata_QC</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">p1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_QC</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Total counts&quot;</span><span class="p">)</span>
<span class="n">p2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_QC</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log1p_norm&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Shifted logarithm&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/50d1833a8d572b79890317f23945ef8d545e18a3ae7c9da35a8d822858cc77cd.png" src="../../_images/50d1833a8d572b79890317f23945ef8d545e18a3ae7c9da35a8d822858cc77cd.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Method-2 : Analytic Pearson residual method ( APR )</span>

</pre></div>
</div>
</div>
</div>
<p><strong>Pearson Residual</strong> :</p>
<p>In pearson residual method the preprocessed counts are compared to the expected counts of a “null model” ( This Null model includes technical variance cells but  no biological variability between cells.)</p>
<p>Pearson residuals :</p>
<ul class="simple">
<li><p>genes that are not differentially expressed = will have variance close to 1.</p></li>
<li><p>gene is differentially expressed, it will deviate from the null model, residual variance &gt;1 for that gene</p></li>
</ul>
<p><strong>&gt; Pearson Residual method helps in :</strong></p>
<ul class="simple">
<li><p>Remove the technical variation that comes from library depth</p></li>
<li><p>stabilize the mean-variance relationship across genes, i.e. ensure that biological signal from both low and high expression genes can contribute similarly to downstream processing</p></li>
<li><p>genes that are homogenously expressed (like housekeeping genes) have small variance, while genes that are differentially expressed (like marker genes) have high variance</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="n">analytic_pearson</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">experimental</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_pearson_residuals</span><span class="p">(</span><span class="n">adata_QC</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">adata_QC</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;APR_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">analytic_pearson</span><span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">])</span>

<span class="n">adata_QC</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;, &#39;APR_counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">p1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_QC</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Total counts&quot;</span><span class="p">)</span>
<span class="n">p2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span>
    <span class="n">adata_QC</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;APR_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;APR_counts&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/b99d1e6176ec38d347ce62c057a765572707267c7bcae465a0ebb38d94828276.png" src="../../_images/b99d1e6176ec38d347ce62c057a765572707267c7bcae465a0ebb38d94828276.png" />
</div>
</div>
<p>As you can see for this data Pearson Residual method performed well .</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_QC</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;, &#39;APR_counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="feature-selection">
<h2>Feature Selection<a class="headerlink" href="#feature-selection" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title select features</span>

<span class="o">%%</span><span class="n">R</span>
<span class="n">library</span><span class="p">(</span><span class="n">scry</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Don/t run this in %%R because it is python code </span>

<span class="n">ro</span><span class="o">.</span><span class="n">globalenv</span><span class="p">[</span><span class="s2">&quot;adata_QC&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_QC</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">sce</span> <span class="o">=</span> <span class="n">devianceFeatureSelection</span><span class="p">(</span><span class="n">adata_QC</span><span class="p">,</span> <span class="n">assay</span><span class="o">=</span><span class="s2">&quot;X&quot;</span><span class="p">)</span>
<span class="n">sce</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 7650 3536 
metadata(13): _scvi_manager_uuid _scvi_uuid ... type_colors umap
assays(5): X ambient_counts counts log1p_norm APR_counts
rownames(7650): AL669831.5 FAM41C ... AL592183.1 AC004556.1
rowData names(14): gene_ids feature_types ... std binomial_deviance
colnames(3536): AGGGTCCCATGACCCG-1-0 ATTCCTAGTGACTGTT-1-0 ...
  GAGGCCTTCTCCTGCA-14-5 CCCTAACAGTTTCTTC-14-5
colData names(23): type sample ... leiden doublet
reducedDimNames(2): PCA UMAP
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">binomial_deviance</span> <span class="o">=</span> <span class="n">ro</span><span class="o">.</span><span class="n">r</span><span class="p">(</span><span class="s2">&quot;rowData(sce)$binomial_deviance&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">idx</span> <span class="o">=</span> <span class="n">binomial_deviance</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[</span><span class="o">-</span><span class="mi">4000</span><span class="p">:]</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">adata_QC</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
<span class="n">mask</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

<span class="n">adata_QC</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_deviant&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mask</span>
<span class="n">adata_QC</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;binomial_deviance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">binomial_deviance</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_QC</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;, &#39;APR_counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<section id="save-qc-fitered-normalized-feature-selected-object">
<h3>Save QC fitered Normalized Feature selected object<a class="headerlink" href="#save-qc-fitered-normalized-feature-selected-object" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">save_file</span> <span class="o">=</span> <span class="s1">&#39;Objects/sc_QCNFS_covid.h5ad&#39;</span>
<span class="n">adata_QC</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">save_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="pca-tsne-umap">
<h2>PCA,tSNE,UMAP<a class="headerlink" href="#pca-tsne-umap" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Load QC Filtered Normalized Feature selected object</span>

<span class="n">adata_QCNFS</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span> <span class="s2">&quot;/content/drive/MyDrive/scRNA_using_Python/Objects/sc_QCNFS_covid.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata_QCNFS</span>

</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;APR_counts&#39;, &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<p><strong>PCA</strong></p>
<p>PCA offers the advantage that it is highly interpretable and computationally efficient. However, as scRNA-seq datasets are rather sparse due to dropout events and therefore highly non-linear, visualization with the linear dimensionality reduction technique PCA is not very appropriate. PCA is typically used to select the top 10-50 PCs which are used for downstream analysis tasks.</p>
<p><strong>t-SNE UMAP</strong></p>
<p>t-distributed stochastic neighbor embedding (t-SNE) as it yielded the best overall performance. Uniform manifold approximation and projection (UMAP) showed the highest stability and separates best the original cell populations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_for_PCA</span> <span class="o">=</span> <span class="n">adata_QCNFS</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">adata_for_PCA</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;APR_counts&#39;, &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="c1"># Perform Z scale Normalization to use for PCA</span>
<span class="c1"># use the count matrix layer zscale_norm_counts</span>

<span class="c1"># USe highly deviance genes as highly variable for PCA</span>

</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="c1"># regress out unwanted variables</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress_out</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">])</span>



<span class="c1"># scale data, clip values exceeding standard deviation 10.</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="n">adata_for_PCA</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;APR_counts&#39;, &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title we use log nomalize data</span>

<span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log1p_norm&quot;</span><span class="p">]</span>

<span class="n">adata_for_PCA</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;APR_counts&#39;, &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title set highl variable as highly deviance </span>

<span class="c1"># setting highly variable as highly deviant to use scanpy &#39;use_highly_variable&#39; argument in sc.pp.pca</span>
<span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_deviant&quot;</span><span class="p">]</span>


</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Perform PCA </span>


<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">,</span> <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_for_PCA</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3536 × 7650
    obs: &#39;type&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;prediction&#39;, &#39;leiden&#39;, &#39;doublet&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;ribo&#39;, &#39;hb&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;n_cells&#39;, &#39;mean&#39;, &#39;std&#39;, &#39;highly_deviant&#39;, &#39;binomial_deviance&#39;, &#39;highly_variable&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;ambient_profile_Gene Expression&#39;, &#39;ambient_profile_all&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;prediction_colors&#39;, &#39;sample_colors&#39;, &#39;type_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;APR_counts&#39;, &#39;ambient_counts&#39;, &#39;counts&#39;, &#39;log1p_norm&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title Plot PCA</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_scatter</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;total_counts&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/e0e9d3fa3f14d278ce1fe8813cbec38430f360753503d3e8dcf5097551a486c5.png" src="../../_images/e0e9d3fa3f14d278ce1fe8813cbec38430f360753503d3e8dcf5097551a486c5.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">components</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;1,2&#39;</span><span class="p">,</span><span class="s1">&#39;3,4&#39;</span><span class="p">,</span><span class="s1">&#39;5,6&#39;</span><span class="p">,</span><span class="s1">&#39;7,8&#39;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/81fea16e68e6b0f48e56dbc285f017a02f450ed21d53702277b88833f57d9b5f.png" src="../../_images/81fea16e68e6b0f48e56dbc285f017a02f450ed21d53702277b88833f57d9b5f.png" />
</div>
</div>
<p>To identify genes that contribute most to each PC, one can retreive the loading matrix information.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#Plot loadings</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_loadings</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">])</span>

<span class="c1"># OBS! only plots the positive axes genes from each PC!!</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/7d87758fbb4416791f79d08ea90a4e89cec33638ede936725623196acb4d0c3f.png" src="../../_images/7d87758fbb4416791f79d08ea90a4e89cec33638ede936725623196acb4d0c3f.png" />
</div>
</div>
<p>The function to plot loading genes only plots genes on the positive axes. Instead plot as a heatmaps, with genes on both postive and negative side, one per pc, and plot their expression amongst cells ordered by their position along the pc.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-825df69f-0b2f-42e2-80cf-67686644a172">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>type</th>
      <th>sample</th>
      <th>batch</th>
      <th>n_genes_by_counts</th>
      <th>total_counts</th>
      <th>pct_counts_in_top_20_genes</th>
      <th>total_counts_mt</th>
      <th>pct_counts_mt</th>
      <th>total_counts_ribo</th>
      <th>pct_counts_ribo</th>
      <th>...</th>
      <th>mt_outlier</th>
      <th>n_counts</th>
      <th>S_score</th>
      <th>G2M_score</th>
      <th>phase</th>
      <th>_scvi_batch</th>
      <th>_scvi_labels</th>
      <th>prediction</th>
      <th>leiden</th>
      <th>doublet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AGGGTCCCATGACCCG-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
      <td>2140</td>
      <td>7698.0</td>
      <td>24.123149</td>
      <td>525.0</td>
      <td>6.819953</td>
      <td>2564.0</td>
      <td>33.307353</td>
      <td>...</td>
      <td>False</td>
      <td>5929.0</td>
      <td>0.115495</td>
      <td>-0.181939</td>
      <td>S</td>
      <td>0</td>
      <td>0</td>
      <td>singlet</td>
      <td>2</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>ATTCCTAGTGACTGTT-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
      <td>1808</td>
      <td>7535.0</td>
      <td>27.458527</td>
      <td>470.0</td>
      <td>6.237558</td>
      <td>3397.0</td>
      <td>45.082946</td>
      <td>...</td>
      <td>False</td>
      <td>5855.0</td>
      <td>0.299016</td>
      <td>0.346879</td>
      <td>G2M</td>
      <td>0</td>
      <td>0</td>
      <td>singlet</td>
      <td>5</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>CCTAAGACAGATTAAG-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
      <td>2449</td>
      <td>9025.0</td>
      <td>19.789474</td>
      <td>513.0</td>
      <td>5.684211</td>
      <td>2637.0</td>
      <td>29.218837</td>
      <td>...</td>
      <td>False</td>
      <td>7611.0</td>
      <td>0.407661</td>
      <td>0.150615</td>
      <td>S</td>
      <td>0</td>
      <td>0</td>
      <td>singlet</td>
      <td>5</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>AATAGAGAGGGTTAGC-1-0</th>
      <td>Covid</td>
      <td>covid_1</td>
      <td>0</td>
      <td>392</td>
      <td>937.0</td>
      <td>38.313767</td>
      <td>63.0</td>
      <td>6.723586</td>
      <td>63.0</td>
      <td>6.723586</td>
      <td>...</td>
      <td>False</td>
      <td>773.0</td>
      <td>0.071607</td>
      <td>-0.092472</td>
      <td>S</td>
      <td>0</td>
      <td>0</td>
      <td>singlet</td>
      <td>8</td>
      <td>singlet</td>
    </tr>
  </tbody>
</table>
<p>4 rows × 23 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-825df69f-0b2f-42e2-80cf-67686644a172')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-825df69f-0b2f-42e2-80cf-67686644a172 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-825df69f-0b2f-42e2-80cf-67686644a172');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  </div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># adata.obsm[&quot;X_pca&quot;] is the embeddings</span>
<span class="c1"># adata.uns[&quot;pca&quot;] is pc variance</span>
<span class="c1"># adata.varm[&#39;PCs&#39;] is the loadings</span>

<span class="n">genes</span> <span class="o">=</span> <span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;gene_ids&#39;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">pc</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]:</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="s1">&#39;PCs&#39;</span><span class="p">][:,</span><span class="n">pc</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">o</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
    <span class="n">sel</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">o</span><span class="p">[:</span><span class="mi">10</span><span class="p">],</span><span class="n">o</span><span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">:]))</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">emb</span> <span class="o">=</span> <span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span><span class="n">pc</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="c1"># order by position on that pc</span>
    <span class="n">tempdata</span> <span class="o">=</span> <span class="n">adata_for_PCA</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">emb</span><span class="p">),]</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">tempdata</span><span class="p">,</span> <span class="n">var_names</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="n">sel</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">swap_axes</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/c7190ea19f09d8f8252d124ad7c3f70e323b544254fc11d43891a04d24cbe4f8.png" src="../../_images/c7190ea19f09d8f8252d124ad7c3f70e323b544254fc11d43891a04d24cbe4f8.png" />
<img alt="../../_images/5305a7b4873082830d641fb406882f604decd425bad2466f646b80014d2f49ae.png" src="../../_images/5305a7b4873082830d641fb406882f604decd425bad2466f646b80014d2f49ae.png" />
<img alt="../../_images/ad8f3610eb06eb67970628cb9a8b5db32c4ca3105160189debe935ecfe1eb973.png" src="../../_images/ad8f3610eb06eb67970628cb9a8b5db32c4ca3105160189debe935ecfe1eb973.png" />
<img alt="../../_images/a223d48cd01a2d09fb402811c9057e1e764d4e413105fb0618bfbbdefec08cb9.png" src="../../_images/a223d48cd01a2d09fb402811c9057e1e764d4e413105fb0618bfbbdefec08cb9.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_pcs</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/3cfd50ade7c1c6d06f12f281f8a3e7a1c539bbabd9491cbc03e335985ac6c99d.png" src="../../_images/3cfd50ade7c1c6d06f12f281f8a3e7a1c539bbabd9491cbc03e335985ac6c99d.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title t-SNE </span>

<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/27ef69307435ab3a5944c2034853b7319ab1a3de897833eae10c79fc923b1cf0.png" src="../../_images/27ef69307435ab3a5944c2034853b7319ab1a3de897833eae10c79fc923b1cf0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#@title U-MAP</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/0852a0f3b6ebb97170f81fec5ace3a72ea4a0e336c7f138f0845dc0b9e7b372b.png" src="../../_images/0852a0f3b6ebb97170f81fec5ace3a72ea4a0e336c7f138f0845dc0b9e7b372b.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#run with 10 components, save to a new object so that the umap with 2D is not overwritten.</span>
<span class="n">umap10</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span><span class="n">constrained_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span>  <span class="n">title</span><span class="o">=</span><span class="s2">&quot;UMAP&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">umap10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;UMAP10&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">components</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1,2&#39;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">umap10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;UMAP10&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">components</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;3,4&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Axes: title={&#39;center&#39;: &#39;UMAP10&#39;}, xlabel=&#39;UMAP3&#39;, ylabel=&#39;UMAP4&#39;&gt;
</pre></div>
</div>
<img alt="../../_images/a9b0a502b8455b542c53c43965d5482a4de2125d55f627a4f49052c3019f952a.png" src="../../_images/a9b0a502b8455b542c53c43965d5482a4de2125d55f627a4f49052c3019f952a.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_for_PCA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD3E&quot;</span><span class="p">,</span> <span class="s2">&quot;CD4&quot;</span><span class="p">,</span> <span class="s2">&quot;CD8A&quot;</span><span class="p">,</span> <span class="s2">&quot;GNLY&quot;</span><span class="p">,</span><span class="s2">&quot;NKG7&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A1&quot;</span><span class="p">,</span><span class="s2">&quot;CD14&quot;</span><span class="p">,</span><span class="s2">&quot;LYZ&quot;</span><span class="p">,</span><span class="s2">&quot;CST3&quot;</span><span class="p">,</span><span class="s2">&quot;MS4A7&quot;</span><span class="p">,</span><span class="s2">&quot;FCGR3A&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/2c57f2aec3962843b36eb6349e1bca874cf8022e3e1c2458f5246664783f53fe.png" src="../../_images/2c57f2aec3962843b36eb6349e1bca874cf8022e3e1c2458f5246664783f53fe.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_for_PCA</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;pct_counts_mt&quot;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_ribo&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_hb&#39;</span><span class="p">,</span> <span class="s2">&quot;doublet&quot;</span><span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/1f782627f38a244d90fc95e38855471ec2753e4880e0a046ed88e2f420605d25.png" src="../../_images/1f782627f38a244d90fc95e38855471ec2753e4880e0a046ed88e2f420605d25.png" />
</div>
</div>
<section id="save-qc-filtered-feature-selected-dm-reduced-data-object">
<h3>Save QC filtered feature selected , DM reduced data object<a class="headerlink" href="#save-qc-filtered-feature-selected-dm-reduced-data-object" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="n">save_file</span> <span class="o">=</span> <span class="s1">&#39;Objects/sc_QCNFSDM_covid.h5ad&#39;</span>
<span class="n">adata_for_PCA</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">save_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./part1\chapter4"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../chapter3/Pre-processing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Pre-processing</p>
      </div>
    </a>
    <a class="right-next"
       href="../chapter5/Data%20Integration.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Data Integration (Batch correction)</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#normalization">Normalization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-selection">Feature Selection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-qc-fitered-normalized-feature-selected-object">Save QC fitered Normalized Feature selected object</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-tsne-umap">PCA,tSNE,UMAP</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-qc-filtered-feature-selected-dm-reduced-data-object">Save QC filtered feature selected , DM reduced data object</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Dr. Majeed Jamakhani
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>